### ISSUE TITLE
s3: Partial file is uploaded when stream errors

### ISSUE THREAD

**Issue by scooper91** [aws/aws-sdk-js#3106]
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js/issues) and didn't find any solution
  
**Describe the bug**

When the stream being uploaded to s3 errors, `upload` errors, but we're finding that the file is still being created (and content partially uploaded).

The partial file does not appear immediately after `s3.upload` errors.

**Is the issue in the browser/Node.js?**
Node.js v12.16.0

**If on Node.js, are you running this on AWS Lambda?**
No.

**SDK version number**
v2.614.0

**To Reproduce (observed behavior)**
```js
const {createReadStream} = require('fs');
const {S3} = require('aws-sdk');
const s3 = new S3();

(async () => {
	const stream = createReadStream('some-large-file');
	setTimeout(() => stream.emit('error', new Error('bang')), 50);

	try {
		await s3.upload({
			Bucket: 'some-bucket',
			Key: 'test',
			Body: stream
		}).promise();
	}
	catch (e) {
		console.log(new Date(), e);

		console.log(new Date(), await s3.listObjectsV2({
			Bucket: 'some-bucket',
			Prefix: 'test',
		}).promise());

		setTimeout(async () => console.log(new Date(), await s3.listObjectsV2({
			Bucket: 'some-bucket',
			Prefix: 'test',
		}).promise()), 100);
	}
})();
```

Output:

```
2020-02-19T17:53:13.469Z Error: bang

2020-02-19T17:53:13.601Z {
  IsTruncated: false,
  Contents: [],
  Name: 'some-bucket',
  Prefix: 'test',
  MaxKeys: 1000,
  CommonPrefixes: [],
  KeyCount: 0
}

2020-02-19T17:53:13.703Z {
  IsTruncated: false,
  Contents: [
    {
      Key: 'test',
      LastModified: 2020-02-19T17:53:14.000Z,
      ETag: '"b8c699b60eef82838f9a905c18064ca4"',
      Size: 402472,
      StorageClass: 'STANDARD'
    }
  ],
  Name: 'some-bucket',
  Prefix: 'test',
  MaxKeys: 1000,
  CommonPrefixes: [],
  KeyCount: 1
}
```

**Expected behavior**

When the upload stream errors, nothing is saved in s3.

**Additional context**

From manually tracing through the SDK, we can see an upload progress event occurring after the stream has errored and `cleanup` has been called.

**Comment by ajredniwja** [comment]
Hey @scooper91 thank-you for reaching out to us with your issue. 

Does this happens every-time using the same code snippet? 

I was not able to reproduce the case using the exact code snippet, is there any more context you can provide. My guess for your use case would be some race condition where the output gets printed before the cleanup is performed.

**Comment by scooper91** [comment]
Hi.

Thanks for the reply.

It is happening every time using that code snippet on my machine. Sorry, the `some-large-file` is misleading - it might be that which is causing the discrepancy in behaviour. Try running `truncate -s 300K some-large-file` to create the file. That is producing consistent results on both my, and a colleague's, machine.

If you still cannot reproduce it, I can try using docker to make it more deterministic.

As far as we can tell, there is a race condition where the upload request is in flight, and then the stream errors. The cleanup doesn't seem to handle that.  Consequently, the file appears in s3 shortly after `s3.upload` errors.

In our actual use case, we're processing a file before uploading to s3, which emits an error when the processing fails.  We're currently having to do a horrible workaround in order to delete the file when that happens.

**Comment by ajredniwja** [comment]
Hey @scooper91, I was using a fairly large file while reproducing it. 

Looking at the code I don't see that the clean-up should fail. 

I am also confused about the 300KB file that you asked to create for the repro.

Can you provide some additional context or a repro repo that I can have a look at?

**Comment by scooper91** [comment]
Hi again @ajredniwja !

I've created a repo [here](https://github.com/scooper91/aws-s3-upload-test). I've tried it on 4 different machines and get the same result each time. Instructions should all be in the [readme](https://github.com/scooper91/aws-s3-upload-test/blob/master/README.md), but let me know if you have any issues.

The output you should see is that the first `listObjectsv2` in the catch returns no items, whereas the second one, after 100ms, returns one item.

**Comment by scooper91** [comment]
@ajredniwja Have you had any luck reproducing this?

**Comment by ajredniwja** [comment]
@scooper91, thank-you for your patience, I worked on it and found it that the race condition is the problem using the code that you provided. 

What happens is that the 300kb file gets uploaded quickly enough that it doesn't hit that 50ms error mark meaning that the file gets uploaded and then the error is emitted and due to the error being emitted late it goes into the catch block and lists the objects. 

I came to that conclusion because I tried it with a relatively large file of 300 mb file which would take more time to upload, if the error is thrown while uploading then it doesn't upload. 

Also I tried the 300kb file with relatively small timeout and it shows the expected behavior.   

Now, if you are running into this issue for your use case, the best workaround would be to do a clean up manually. 

I can help you with your use case if you have any additional questions.
